drain/drain.py

```diff
@@ -0,0 +1,417 @@
+    if check_state:
+        if channel['state'] != "CHANNELD_NORMAL":
+            raise RpcError(payload['command'], payload, {'message': 'Channel %s not in state CHANNELD_NORMAL, but: %s' % (scid, channel['state']) })
+        if not peer['connected']:
+            raise RpcError(payload['command'], payload, {'message': 'Channel %s peer is not connected.' % scid})
+    return channel
+
+
+def spendable_from_scid(plugin, payload, scid):
+    # only fetch funds once to reduce RPC load
+    if not "funds" in payload:
+        payload['funds'] = plugin.rpc.listfunds().get('channels')
+
+    try:
+        channel_funds = next(c for c in payload['funds'] if 'short_channel_id' in c and c['short_channel_id'] == scid)
+    except StopIteration:
+        return Millisatoshi(0), Millisatoshi(0)
+    peer_id = channel_funds['peer_id']
+    funds_our = Millisatoshi(channel_funds['our_amount_msat'])
+    try:
+        channel_peer = get_channel(plugin, payload, peer_id, scid, True)
+    except RpcError:
+        return Millisatoshi(0), Millisatoshi(0)
+
+    # we check amounts via gossip and not wallet funds, as its more accurate
+    our = Millisatoshi(channel_peer['to_us_msat'])
+    total = Millisatoshi(channel_peer['total_msat'])
+    our_reserve = Millisatoshi(channel_peer['our_reserve_msat'])
+    their_reserve = Millisatoshi(channel_peer['their_reserve_msat'])
+    their = total - our
+
+    # reserves maybe not filled up yet
+    if our < our_reserve:
+        our_reserve = our
+    if their < their_reserve:
+        their_reserve = their
+
+    spendable = channel_peer['spendable_msat']
+    receivable = their - their_reserve
+    return spendable, receivable
+
+
+def peer_from_scid(plugin, payload, short_channel_id, my_id):
+    channels = plugin.rpc.listchannels(short_channel_id).get('channels')
+    for ch in channels:
+        if ch['source'] == my_id:
+            return ch['destination']
+    raise RpcError(payload['command'], payload, {'message': 'Cannot find peer for channel: ' + short_channel_id})
+
+
+def find_worst_channel(route):
+    if len(route) < 4:
+        return None
+    start_id = 2
+    worst = route[start_id]['channel']
+    worst_val = route[start_id - 1]['msatoshi'] - route[start_id]['msatoshi']
+    for i in range(start_id + 1, len(route) - 1):
+        val = route[i - 1]['msatoshi'] - route[i]['msatoshi']
+        if val > worst_val:
+            worst = route[i]['channel']
+            worst_val = val
+    return worst
+
+
+def test_or_set_chunks(plugin, payload, my_id):
+    scid = payload['scid']
+    cmd = payload['command']
+    spendable, receivable = spendable_from_scid(plugin, payload, scid)
+    total = spendable + receivable
+    target = total * (0.01 * (100 - payload['percentage']))
+
+    if cmd == "drain":
+        if target >= spendable:
+            raise RpcError(payload['command'], payload, {'message': 'channel already below target value'})
+        amount = spendable - target
```

Comment:
Trying this some more I think I don't get the logic (or don't understand the params) :
I tried to drain some sats from a channel with `293600000`msat (`287728000` spendable), setting the percentage to 30 I effectively drained `33807374`msat, which is about 11%of the channel capacity:
```python
>>> 33807374 / 287728000 * 100
11.749768531390758
>>> 33807374 / 293600000 * 100
11.514773160762942
```
Hmm trying this again I then got an error that the target was above the capacity (or below to fill), even with 1% as parameter ! Is this the error you were talking about ?

Back to the highlighted lines I think you compute the target as the final channel capacity:
```python
target = total * (0.01 * (100 - payload['percentage']))
```
Then you compare it to the spendable amount
```python
if target >= spendable:
```
Which will likely be inferior.
I think that the target should be "how much do I want to drain" and thus the target be :
```python
target = total * payload['percentage'] / 100
```
Or 
```python
    # The amount to either fill or drain
    amount = total * payload['percentage'] / 100

    if cmd == "drain":
        if amount >= spendable:
             # Hmm problem here
    elif cmd == "fill":
        if amount >= receivable:
             # Hmmmm
```

But maybe I'm just misunderstanding and misusing the whole thing

Reply:
Yes, the percentage amount is the resulting final channel capacity. The error can be reproduced when quickly draining and filling several channels (i.e. 100% default) until the error is raised.

Reply:
Note: I edited my last response to be more precise.

Reply:
After more testing I think that's the latter case.. :
> But maybe I'm just misunderstanding and misusing the whole thing

Reply:
But I still don't understand why I could not drain (or fill) 1% of a channel that is completely balanced (nor the target logic)..

Reply:
That's easy, a balanced channel is by definition 50% filled and drained 'at the same time'. You cannot be able to 'drain' to 1% as that would equate to a fill 99% (but command inversion is a Todo and not implemented yet.). Equally a fill to 1% is impossible as it has already 50% which is greater than 1%.

The percentage do not related to current balance but to target amount.

Reply:
Do you think we should implement command handling differently, if yes how exactly?

Current pros:
- do not need to know current percentage l
- no amount calculation needed
- Able to rebalance by using drain or fill 50%

Cons:
- counter intuitive? :D
- 'not often possible' to do stuff like drain 1% as that would require to have the channel above 99%.

Reply:
Thanks for the explanation !

> Do you think we should implement command handling differently, if yes how exactly?

It was counter intuitive *to me*, maybe because I haven't read "mastering english" yet :smile: 
However I'd argue to maybe change the command name ? According to my understanding, `drain <amount> <scid>` implies `I want to drain <amount>% of sats from <scid>` : actually I tried to use it as a rebalance. Maybe `set_ours <percent>` `set_them <percent>`, they can still be additional commands though.

>  I think from a UX requirements perspective it is important to be able to get a channel balance full or empty (hence 100% default) or some to fractional value without doing math calculations. It is not a useful feature to send 1% or X% of a channel elsewhere.

I totally agree. Just asking as many questions as possible to make this as perfect as possible ;-) Next step (once merged) is the htlc fee finding optimisation

Reply:
@darosior can you come to irc? Or can we have a quick talk somewhere?

Reply:
Yep I pinged you on IRC

Reply:
> How about this:

    drain X will send X% total (or current?) channel balance away to some other channel, if possible.
    fill X will receive X% total (or current?) channel balance from some other channel, if possible.
    reach X will try to achieve X% our capacity, if possible. This command will actually do the auto conversion (drain/fill) if channel is above or below.

Sounds great

Reply:
Ok so I have a channel with 60% theirs, 40% ours. I think the wording is misleading as if I want to drain, say 30 percent of the channel (so that the balance results in 90% theirs, 10% ours), I dont
```
cli drain <scid> 30
```
But
```
cli drain <scid> 90
```
So effectively passing to the drain command the expected percent on `their` side. That's why I proposed to change the wording to :
```
cli set_theirs <scid> 90
```
And, if we keep the `drain` command, to pass it the expected percent to __drain from__ my side.

The same goes for `fill` :
Now that they effectively have 90% of the channel capacity, and I want my side to have 90% capacity (so passing from `ours: 10%; theirs: 90%` to `ours: 90%; theirs: 10%`) I would expect to drain __from them__ 80% of the capacity (hence `fill` 80% on my side) but in order to do that with the current wording I have to :
```
cli fill <scid> 90
```
That's why I propose to rename it to 
```
cli set_ours <scid> 80
```

Reply:
will be address by implementing this TODO 8d555a46dbd6dba32485ce4ea5a78955bae224ff